mapiproxy/libmapistore/mapistore_notification.c

```diff
@@ -3,525 +3,1355 @@
+	} else {
+		notification_ctx->memc_ctx = (memcached_st *) memcached_create(NULL);
+		MAPISTORE_RETVAL_IF(!notification_ctx->memc_ctx, MAPISTORE_ERR_CONTEXT_FAILED, notification_ctx);
+
+		servers = memcached_server_list_append(servers, MSTORE_MEMC_DFLT_HOST, 0, &rc);
+		MAPISTORE_RETVAL_IF(!servers, MAPISTORE_ERR_CONTEXT_FAILED, notification_ctx);
+
+		rc = memcached_server_push(notification_ctx->memc_ctx, servers);
+		memcached_server_list_free(servers);
+		MAPISTORE_RETVAL_IF(rc != MEMCACHED_SUCCESS, MAPISTORE_ERR_CONTEXT_FAILED, notification_ctx);
 	}
 
-        return new_subscription;
-#endif
+	talloc_set_destructor((void *)notification_ctx, (int (*)(void *))mapistore_notification_destructor);
 
-	return NULL;
+	*_notification_ctx = notification_ctx;
+	return MAPISTORE_SUCCESS;
 }
 
-_PUBLIC_ void mapistore_push_notification(struct mapistore_context *mstore_ctx, uint8_t object_type, enum mapistore_notification_type event, void *parameters)
+
+/**
+   \details Generate the session key
+
+   \param mem_ctx pointer to the memory context
+   \param async_uuid asynchronous emsmdb session uuid to compute
+   \param _key pointer on pointer to the key to return
+
+   \note the caller is responsible for freeing _key
+
+   \return MAPISTORE_SUCCES on success, otherwise MAPISTORE error
+ */
+static enum mapistore_error mapistore_notification_session_set_key(TALLOC_CTX *mem_ctx,
+								   struct GUID async_uuid,
+								   char **_key)
+{
+	char	*guid = NULL;
+	char	*key = NULL;
+
+	/* Sanity checks */
+	MAPISTORE_RETVAL_IF(!_key, MAPISTORE_ERR_INVALID_PARAMETER, NULL);
+
+	guid = GUID_string(mem_ctx, (const struct GUID *)&async_uuid);
+	MAPISTORE_RETVAL_IF(!guid, MAPISTORE_ERR_NO_MEMORY, NULL);
+
+	key = talloc_asprintf(mem_ctx, MSTORE_MEMC_FMT_SESSION, guid);
+	talloc_free(guid);
+	MAPISTORE_RETVAL_IF(!key, MAPISTORE_ERR_NO_MEMORY, NULL);
+
+	*_key = key;
+	return MAPISTORE_SUCCESS;
+}
+
+
+/**
+   \details Register and bind asynchronous emsmdb session handle to
+   emsmdb one
+
+   \param mstore_ctx pointer to the mapistore context
+   \param uuid the emsmdb session uuid
+   \param async_uuid the asynchronous emsmdb session uuid
+   \param cn the common name (username or email) to associate
+
+   \return MAPISTORE_SUCCESS on success, otherwise MAPISTORE error
+ */
+_PUBLIC_ enum mapistore_error mapistore_notification_session_add(struct mapistore_context *mstore_ctx,
+								 struct GUID uuid,
+								 struct GUID async_uuid,
+								 const char *cn)
+{
+	TALLOC_CTX				*mem_ctx;
+	enum mapistore_error			retval;
+	struct mapistore_notification_session	r;
+	struct ndr_push				*ndr;
+	enum ndr_err_code			ndr_err_code;
+	char					*key;
+	memcached_return			rc;
+
+	/* Sanity checks */
+	MAPISTORE_RETVAL_IF(!mstore_ctx, MAPISTORE_ERR_NOT_INITIALIZED, NULL);
+	MAPISTORE_RETVAL_IF(!mstore_ctx->notification_ctx, MAPISTORE_ERR_NOT_AVAILABLE, NULL);
+	MAPISTORE_RETVAL_IF(!mstore_ctx->notification_ctx->memc_ctx, MAPISTORE_ERR_NOT_AVAILABLE, NULL);
+
```

Comment:
The original reason was to test what would be the ndr_ layer behavior if a malicious user was forging a notification with empty cn. I wanted to guarantee an error would be triggered instead of a crash - which is the current behavior.